File: app/api/arxiv/route.ts
import { NextResponse } from "next/server";
import fetch from 'node-fetch'; // Ensure you have node-fetch installed
import { XMLParser } from 'fast-xml-parser'; // Import the XML parser

export async function POST(req) {
    const { topic, maxResults } = await req.json();

    const url = `http://export.arxiv.org/api/query?search_query=ti:"${encodeURIComponent(topic)}"&sortBy=relevance&sortOrder=ascending&max_results=${maxResults}`;

    try {
        const response = await fetch(url);
        const text = await response.text();

        // Parse the XML response using fast-xml-parser
        const parser = new XMLParser();
        const jsonObj = parser.parse(text);

        // Check if entries exist
        if (!jsonObj.feed.entry) {
            return NextResponse.json({ results: [] }); // Return empty array if no entries
        }

        // Extract and format the entries
        const entries = Array.isArray(jsonObj.feed.entry) ? jsonObj.feed.entry : [jsonObj.feed.entry];

        return NextResponse.json(entries);
    } catch (error) {
        console.error('Error fetching data from arXiv:', error);
        return NextResponse.json({ error: 'Failed to fetch data from arXiv' }, { status: 500 });
    }
}

File: app/api/arxiv/python/route.py
from fastapi import FastAPI, HTTPException
from pydantic import BaseModel
import requests
import xmltodict

app = FastAPI()

class QueryRequest(BaseModel):
    topic: str
    maxResults: int

@app.post("/fetch_arxiv/")
async def fetch_arxiv(request: QueryRequest):
    topic = request.topic
    max_results = request.maxResults

    url = f"http://export.arxiv.org/api/query?search_query=ti:{topic}&sortBy=relevance&sortOrder=ascending&max_results={max_results}"
    
    print("ARXIV")
    print(url)

    try:
        response = requests.get(url)
        response.raise_for_status()  # Raise an error for bad responses
        data = xmltodict.parse(response.text)

        # Check if 'entry' exists in parsed XML
        if "feed" not in data or "entry" not in data["feed"]:
            return {"results": []}  # Return an empty array if no entries

        # Ensure entries are always a list
        entries = data["feed"]["entry"]
        if not isinstance(entries, list):
            entries = [entries]

        return {"results": entries}

    except requests.RequestException as e:
        print(f"Error fetching data from arXiv: {e}")
        raise HTTPException(status_code=500, detail="Failed to fetch data from arXiv")
